core/arch/arm/sm/std_smc.c

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (C) 2016 Freescale Semiconductor, Inc.
+ * All rights reserved.
+ *
+ * Peng Fan <peng.fan@nxp.com>
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <stdint.h>
+#include <sm/optee_smc.h>
+#include <sm/psci.h>
+#include <sm/sm.h>
+#include <sm/std_smc.h>
+#include <tee/uuid.h>
+#include <trace.h>
+
+TEE_UUID uuid = {
+	0x108d905b, 0xf863, 0x47e8,
+	{0xae, 0x2d, 0xc0, 0xfb, 0x56, 0x41, 0xf6, 0xe2},
+};
+
+bool smc_std_handler(struct thread_smc_args *args)
+{
+	uint32_t smc_fid = args->a0;
+
+	if (OPTEE_SMC_OWNER_NUM(smc_fid) != OPTEE_SMC_OWNER_STANDARD)
```

Comment:
This test should preferably be done in `sm_from_nsec()`, before calling this function as it would allow the compiler to optimize `sm_from_nsec()` better.

Something like (but perhaps using yet another wrapper as PSCI support is supposed to be configurable):
```
if (OPTEE_SMC_OWNER_NUM(ctx->nsec.r0) == OPTEE_SMC_OWNER_STANDARD) {
          smc_std_handler((struct thread_smc_args *)&ctx->nsec.r0);
          return true;
}
```